llama : greatly reduce output buffer memory usage

compilade opened the pull request:
Supersedes #2700.

As I've noted in <https://github.com/ggerganov/llama.cpp/pull/6017#discussion_r1523483648>, the logits buffer is way too big and mostly unused when the batch size is very large. This PR fixes this waste of memory by introducing another, smaller buffer of ids (`ctx->output_ids`) which point into the logits and embeddings buffers, allowing to keep most of the behavior of `llama_get_logits_ith()` and `llama_get_embeddings_ith()` while making the output buffer's content contiguous. While I was at it, I've also noticed it was relatively easy to skip computing unused logits with the new output buffer layout introduced in this change.

From @slaren's suggestion in <https://github.com/ggerganov/llama.cpp/pull/6017#discussion_r1523530117>, this PR allocates space for `n_seq_max` outputs, then dynamically reallocates the output buffer when more logits and/or embeddings are necessary.
(Thanks for making me think to look for #2700 when alluding to it in <https://github.com/ggerganov/llama.cpp/pull/6017#discussion_r1523579895>)

# API changes

- `llama_get_logits` and `llama_get_embeddings` now return contiguous arrays (still `float *`).
  - The logits and/or the embeddings for which `llama_batch.logits[i] == true` are stored contiguously in the order they have in the batch.
  - The layout is the same as before if `logits_all == true` and when using `llama_batch_get_one()` (which doesn't set `llama_batch.logits`).
  - Only the necessary logits and embeddings are computed, so reducing the use of `logits_all` now has a *slight* performance incentive
- `llama_get_logits_ith` and `llama_get_embeddings_ith` now always verify if the passed id is valid.
  - The result was undefined in some cases before.
  - Note that the `perplexity` example previously relied on the non-verification of this (assertion error when running with a Debug build on `master`), which has been fixed. (in other words, this fixes #6246)
- The session file format changed slightly, and is now **much smaller** (a lot of space for unused logits was wasted before). :tada:
  - For example, storing the state of Tinyllama with a batch size of `512` previously used at least `63 MiB`, but now the equivalent session file takes only `1 MiB` (mostly composed of the used KV cache cells from my test prompt, it now only includes the needed logits from the last batch), regardless of the batch size, whereas a batch size of `1024` previously would use around `126 MiB` with Tinyllama.
  - It's now possible to load a session file with a different context size than what it was saved with, as long as the new context size is at least enough to contain the state for the saved tokens.

# Notes

The `perplexity` example used the previous layout of the logits very extensively.
I've adapted most of it, but the parts of it which still use `logits_all` (even though this still works) will need to be changed in the future to benefit from skipping the computation of unused logits.

The `perplexity` example should have the exact same output as `master` when compiled with the same flags. ~~Well, except with Winogrande, because I've modified the implementation.~~ (EDIT: Winogrande output should be the same as `master` as of 8f70dcb0f3b40cf333a1652e6ebab16640636edd.)
(note to @ikawrakow: the logic for skipping the choice words in `winogrande_score()` shouldn't be required, but I still kept it. ~~so I've simplified it by also fixing the logits used when evaluating the second choice (previously, the logits of the end of the first choice were used there instead of the logits of the end of the common prefix, which caused a big skew in the log-likelyhood of some second choices)~~. This will need to be fixed in a separate PR.)

# TODO

Since a (small) model-specific change was required for each of the 23+ architectures supported by `llama.cpp`, I'd like to at least ensure I didn't break any of them. I'd really like to know if it works on GPU and/or with MoE models.

**Feel free to edit the list below to add more tested models and examples.**

Compare the output with `--temp 0` when using this PR vs `master`.

- [x] Mamba (on CPU: `main`, `parallel` (with and without `-cb`), `perplexity` (v1, v2, hellaswag, multiple-choices), `imatrix`, `save-load-state`, `embedding`)
- [x] LLama (with [Tinyllama](https://huggingface.co/TheBloke/TinyLlama-1.1B-Chat-v1.0-GGUF) (on CPU: `speculative` (with [Llama-160M](https://huggingface.co/Felladrin/gguf-Llama-160M-Chat-v1) as a draft model), `save-load-state`))
  - [x] EDIT (2024-03-19): seems to differ now, investigating... False alarm, it's because repetition penalty is now disabled by default in `master` since <https://github.com/ggerganov/llama.cpp/pull/6127>, so for the outputs to be the same, `--repeat-penalty 1` now has to be passed.
- [x] Phi-2 (on CPU: `main`, `save-load-state`)
- [x] GritLM
- [x] BERT (from the CI test-suite: `server`)
- [x] Any Moe Model (Mixtral ([with HipBLAS](https://github.com/ggerganov/llama.cpp/pull/6122#issuecomment-2002995233): `llama-bench`, `main`))
- [ ] (17 other model types (TODO: expand this list))

Known issues:
- [x] ~~`server` with embeddings~~
- [ ] Some backends don't support `GGML_OP_GET_ROWS` and don't fallback properly, (e.g. the Vulkan backend)

compilade committed:
llama : greatly reduce logits memory usage

compilade committed:
llama : more compact state saving and reloading

compilade committed:
llama : fix lctx.n_outputs not being set before building graph

compilade committed:
perplexity : adapt to the logits API changes

compilade committed:
perplexity : fix Winogrande, use correct logits for second choice start

The first logits used to evaluate the second choice were not from
the end of the common prefix; instead, they were the logits from the end
of the first choice. This has been corrected.

The previous implementation sometimes had outliers in the scores of
choices for some tasks, and the logic to skip choices words
in the log-likelihood evaluation probably was an attempt to reduce those,
but it was complex and didn't quite seem to be the right thing.

This is simpler now, and the outlier scores aren't there anymore.

compilade committed:
perplexity : normalize spaces and punctuation in Winogrande sentences

compilade committed:
llama : fix embedding conditions

compilade committed:
llama : fix llama_get_embeddings_ith when the resulting id is 0

compilade committed:
llama : fix wrong n_outputs in llama_set_inputs

A mismatch happened when using a smaller n_ubatch than n_batch and then using
llama_batch_get_one(). The decision of what n_outputs should be now almost
fully depends on how lctx.n_outputs is set in llama_decode_internal.
The conditions are simpler this way.

* llama : when saving the state, recalculate n_outputs

This ensures the correct number of outputs for the entire previous batch
is stored in the session file, even when n_ubatch is smaller than n_batch.

slaren commented on llama.cpp:
```diff
@@ -8773,9 +8978,39 @@ static void llama_set_inputs(llama_context & lctx, const llama_batch & batch) {
     }
 
+    {
+        GGML_ASSERT(lctx.inp_out_ids && "every model type must skip unused outputs");
```
`inp_out_ids` is skipped when there are no outputs, so this assert is not correct. `inp_out_ids` should be set to `nullptr` in `llm_build_context::init`.

compilade replied:
Thanks, should be fixed now.

slaren replied:
This is not enough to fix it, as it is possible to have batches without outputs. For example, when evaluating a prompt larger than `n_batch`, the first batches will have no outputs. However this may not be relevant after fixing the pipeline parallelism issue.

compilade replied:
I understand the problem, but what's weird is that I didn't run into it before, even when I tried. It was probably because `lctx.inp_out_ids` was initialized to a garbage address and it somehow still worked.

compilade committed:
llama : fix not-skipping outputs of non-causal models

slaren commented:
As it is, this breaks pipeline parallelism because changes in the graph topology force a synchronization. I think it should be possible to fix this if the final `get_rows` is done unconditionally.

You can test this without multiple GPUs by building in debug with `LLAMA_DEBUG`, and using `llama-bench` with `n_batch > n_ubatch`, or `perplexity` with multple sequences. If you get `ggml_backend_sched_alloc_splits: failed to allocate graph, reserving` messages on every eval, then it is breaking pipeline parallelism.

compilade committed:
llama : fix running a batch with n_outputs == 0

It previously worked because lctx.inp_out_ids was not initialized,
so it pointed to some garbage address which was somehow still valid when I
ran my tests.

compilade commented:
> As it is, this breaks pipeline parallelism because changes in the graph topology force a synchronization. I think it should be possible to fix this if the final `get_rows` is done unconditionally.

Yes, this should be possible. I initially put the skip of the rest of the graph there because tensors with 0 rows caused division by zero problems in `ggml_can_repeat`. But working around this could allow keeping the same graphs (assuming no other problems, like *more* divisions by zero elsewhere).

> You can test this without multiple GPUs by building in debug with `LLAMA_DEBUG`, and using `llama-bench` with `n_batch > n_ubatch`, or `perplexity` with multple sequences. If you get `ggml_backend_sched_alloc_splits: failed to allocate graph, reserving` messages on every eval, then it is breaking pipeline parallelism.

I'll see if I can try this on my Intel UHD Graphics 615 (since splits don't seem to happen on CPU-only).
But on CPU, when using the `parallel` example (and also `perplexity` with multiple sequences), I do see a bunch of
```
ggml_gallocr_needs_realloc: graph has different number of nodes
ggml_gallocr_alloc_graph: reallocating buffers automatically
```
which may be problematic when using a GPU.

I'll try to avoid changing the graph topology.

slaren commented:
> I'll see if I can try this on my Intel UHD Graphics 615 (since splits don't seem to happen on CPU-only).

Yes, I forgot that the reallocation is automatic if there is only one backend, but that message also shows the problem. If that message disappears, then it should also work with pipeline parallelism.

compilade committed:
llama : keep same graph topology even when n_outputs == 0

compilade committed:
ggml : saner ggml_can_repeat with empty tensors

*  ggml : future-proof ggml_is_empty by using GGML_MAX_DIMS - 1

compilade committed:
ggml : do not multi-thread ops returning empty tensors

fgdfgfthgr-fox commented:
Test result using Radeon VII HipBLAS build under Linux environment.
I don't see any significant decrease in VRAM use though. In what case should this pr 's change to be significant?
master:
```
$ ./llama-bench -m /mnt/2878EBCCAED823C6/koboldcpp-rocm/mixtral/mixtral-8x7b-iq3_xxs.gguf -ngl 22
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 ROCm devices:
  Device 0: AMD Radeon VII, compute capability 9.0, VMM: no
| model                          |       size |     params | backend    | ngl | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------- | ---------------: |
| llama 7B IQ3_XXS - 3.0625 bpw  |  16.99 GiB |    46.70 B | ROCm       |  22 | pp 512     |     87.74 ± 0.11 |
| llama 7B IQ3_XXS - 3.0625 bpw  |  16.99 GiB |    46.70 B | ROCm       |  22 | tg 128     |      7.55 ± 0.21 |

build: d01b3c4c (2450)
Peak VRAM Used: 13376MB

$ ./main -ngl 22 -m /mnt/2878EBCCAED823C6/koboldcpp-rocm/mixtral/mixtral-8x7b-iq3_xxs.gguf -n 64 --temp 0  -p "In a land before time"
...
generate: n_ctx = 512, n_batch = 2048, n_predict = 64, n_keep = 1


 In a land before time, when the internet was still in its infancy and social media didn’t exist, there were only two ways to get your message out: print or broadcast.

Print is great for reaching people who are already interested in what you have to say – but it can be expensive and hard to measure results. Broad
llama_print_timings:        load time =    2615.59 ms
llama_print_timings:      sample time =       7.09 ms /    64 runs   (    0.11 ms per token,  9026.80 tokens per second)
llama_print_timings: prompt eval time =     715.49 ms /     6 tokens (  119.25 ms per token,     8.39 tokens per second)
llama_print_timings:        eval time =    8572.94 ms /    63 runs   (  136.08 ms per token,     7.35 tokens per second)
llama_print_timings:       total time =    9315.96 ms /    69 tokens
```
pr:
```
$ ./llama-bench -m /mnt/2878EBCCAED823C6/koboldcpp-rocm/mixtral/mixtral-8x7b-iq3_xxs.gguf -ngl 22
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 ROCm devices:
  Device 0: AMD Radeon VII, compute capability 9.0, VMM: no
| model                          |       size |     params | backend    | ngl | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------- | ---------------: |
| llama 7B IQ3_XXS - 3.0625 bpw  |  16.99 GiB |    46.70 B | ROCm       |  22 | pp 512     |     90.56 ± 0.13 |
| llama 7B IQ3_XXS - 3.0625 bpw  |  16.99 GiB |    46.70 B | ROCm       |  22 | tg 128     |      7.41 ± 0.13 |

build: 6bf7f3f4 (2464)
Peak VRAM Used: 13325MB

$ ./main -ngl 22 -m /mnt/2878EBCCAED823C6/koboldcpp-rocm/mixtral/mixtral-8x7b-iq3_xxs.gguf -n 64 --temp 0  -p "In a land before time"
...
generate: n_ctx = 512, n_batch = 2048, n_predict = 64, n_keep = 1


 In a land before time, when the internet was still in its infancy and social media didn’t exist, there were only two ways to get your message out: print or broadcast.

Print is great for reaching people who are already interested in what you have to say – but it can be expensive and hard to measure results. Broad
llama_print_timings:        load time =    2659.17 ms
llama_print_timings:      sample time =       7.00 ms /    64 runs   (    0.11 ms per token,  9148.08 tokens per second)
llama_print_timings: prompt eval time =     551.66 ms /     6 tokens (   91.94 ms per token,    10.88 tokens per second)
llama_print_timings:        eval time =    8741.20 ms /    63 runs   (  138.75 ms per token,     7.21 tokens per second)
llama_print_timings:       total time =    9320.75 ms /    69 tokens
```

slaren commented on ggml.c:
```diff
@@ -2542,6 +2542,12 @@ static inline bool ggml_is_padded_1d(const struct ggml_tensor * tensor) {
 }
 
+static inline bool ggml_is_empty(const struct ggml_tensor * tensor) {
+    // the stride of the last dimension is zero if any of the previous dimensions has no elements
+    // but to be sure, the number of elements in the last dimension must also be checked
+    return tensor->nb[GGML_MAX_DIMS - 1] == 0 || tensor->ne[GGML_MAX_DIMS - 1] == 0;
```
I think it is good to add support for empty tensors, it will allow writing code with fewer special cases, but it will require updating all the backends to skip empty ops. To facilitate that, this function should be exposed in `ggml.h`.

This implementation will not work reliably with views, do something like this instead:

```suggestion
    for (int i = 0; i < GGML_MAX_DIMS; ++i) {
        if (tensor->ne[i] == 0) {
            return true;
        }
    }
    return false;
```

slaren commented on llama.cpp:
```diff
@@ -2094,19 +2097,19 @@ struct llama_context {
+    size_t    logits_size = 0; // capacity (of floats) for logits
+    float   * logits      = nullptr;
+
+    int32_t * output_ids  = nullptr; // map token positions to ids of the logits and embd buffers
```
Use a `std::vector` instead.

compilade committed:
ggml : make ggml_is_empty public and work with views

compilade committed:
llama : use a vector for ctx->output_ids

* llama : rework reallocation logic for llama_output_reserve

Now comparing the actual size with the new total size of the output buffer
to allow more efficient enabling and disabling of the embeddings
and/or logits output in the future.

compilade committed:
ggml : skip empty tensors in all backends

compilade committed:
llama : fix llama_output_reserve nullptr deref when new_size is 0

compilade commented:
Answering 2 messages at once:

> I don't see any significant decrease in VRAM use though.

@fgdfgfthgr-fox The expected decrease is at most `n_vocab*(n_batch - 1)*sizeof(float)` bytes, so for Mixtral, if its `n_vocab` is around 50000 and if using `n_batch == 512`, the maximum decrease will be 97 MiB, while with `llama-bench`, you're seeing a decrease of `50 MiB`, which seems reasonable. BTW, thanks for testing this with Mixtral :)

> In what case should this pr 's change to be significant?

It should be significant when using a very large batch size. For example, with `-b 8192`, this PR could save up to `1.5 GiB` of memory (or `3.05 GiB` of memory with `-b 16384`, etc.) compared to `master`.
With `llama-bench`, I think this can be tested by adding the flags `-b 8192 -p 8192`, while with `main`, `-b 8192 -c 8192` could be used (to ensure the batch size won't be clamped by the context size). Of course, if the KV cache (unaffected by this PR) is too big for your system, you can use smaller numbers.

---

> That sounds really promising! However, when trying to run this PR, I'm getting this error:

@Dampfinchen Thanks for testing this with CUDA. A more complete backtrace would be helpful[^1], but in this case this is probably happening because I didn't yet make the other backends than CPU skip empty tensors, so what you're seeing is likely the symptoms of a GPU-accelerated division by zero. Hopefully fixed by 8b826c5b086c94aaf09977bf41cfbf7ae32786ed. This didn't happen to @fgdfgfthgr-fox because they didn't offload enough layers for the last one to be offloaded (which can now have tensors with no elements when no logits are used, causing problems when dividing dimensions with each other if not skipped).

[^1]: Apparently, `GGML_ASSERT` prints full backtraces when `gdb` is installed on Linux and Unix-like systems. I didn't know for a long time because the only debugger I had already installed was `lldb`. Knowing this (or `coredumpctl debug --debugger=lldb`, followed by `bt`) two months ago would have saved me a lot of `printf`-based debugging.

> Does this PR only support llama or its derivatives like Mistral as well?

The goal is for this to support every of the 23+ model architectures supported by `llama.cpp`. In theory, they should all work, but maybe there's something I overlooked somewhere[^2], which is why I encourage testing models of different types and different backends. Mistral should work, and from <https://github.com/ggerganov/llama.cpp/pull/6122#issuecomment-2002995233>, Mixtral too.

[^2]: like the Vulkan backend not properly supporting `GGML_GET_ROWS`... Probably wasn't too important before because `ggml_get_rows` was mostly used on input tensors which are always using the CPU backend. (but it was also used for MoE inference, so backends on which Mixtral worked likely do not have this problem)

slaren commented:
Works well with CUDA, improves pp performance by 2-3% with a single GPU.

| GPU         | Model         |   Model Size [GiB] | Test   |   t/s master |   t/s PR |   Speedup |
|:------------|:--------------|-------------------:|:-------|-------------:|--------------------------------------:|----------:|
| RTX 3090 Ti | llama 7B F16  |              12.55 | pp1024 |      5031.00 |                               5144.94 |      1.02 |
| RTX 3090 Ti | llama 7B F16  |              12.55 | pp2048 |      4600.25 |                               4710.85 |      1.02 |
| RTX 3090 Ti | llama 7B F16  |              12.55 | pp4096 |      3956.06 |                               4032.93 |      1.02 |
| RTX 3090 Ti | llama 7B Q4_0 |               3.56 | pp1024 |      4190.07 |                               4316.03 |      1.03 |
| RTX 3090 Ti | llama 7B Q4_0 |               3.56 | pp2048 |      3904.58 |                               4011.76 |      1.03 |
| RTX 3090 Ti | llama 7B Q4_0 |               3.56 | pp4096 |      3429.24 |                               3512.89 |      1.02 |
| RTX 3090 Ti | llama 7B F16  |              12.55 | tg128  |        54.76 |                                 54.81 |      1.00 |
| RTX 3090 Ti | llama 7B Q4_0 |               3.56 | tg128  |       126.81 |                                126.19 |      1.00 |

| GPU                  | Model         |   Model Size [GiB] | Test   |   t/s master |   t/s PR |   Speedup |
|:---------------------|:--------------|-------------------:|:-------|-------------:|--------------------------------------:|----------:|
| RTX 3090 Ti/RTX 3080 | llama 7B F16  |              12.55 | pp1024 |      5724.30 |                               5759.07 |      1.01 |
| RTX 3090 Ti/RTX 3080 | llama 7B F16  |              12.55 | pp2048 |      6146.06 |                               6129.15 |      1.00 |
| RTX 3090 Ti/RTX 3080 | llama 7B F16  |              12.55 | pp4096 |      5783.30 |                               5803.46 |      1.00 |
| RTX 3090 Ti/RTX 3080 | llama 7B Q4_0 |               3.56 | pp1024 |      4825.07 |                               4909.32 |      1.02 |
| RTX 3090 Ti/RTX 3080 | llama 7B Q4_0 |               3.56 | pp2048 |      5275.33 |                               5319.94 |      1.01 |
| RTX 3090 Ti/RTX 3080 | llama 7B Q4_0 |               3.56 | pp4096 |      5026.15 |                               5091.23 |      1.01 |
| RTX 3090 Ti/RTX 3080 | llama 7B F16  |              12.55 | tg128  |        48.34 |                                 47.83 |      0.99 |
| RTX 3090 Ti/RTX 3080 | llama 7B Q4_0 |               3.56 | tg128  |       108.05 |                                108.19 |      1.00 |

slaren commented:
Note that the output buffer is always allocated in a CPU buffer, so this shouldn't affect VRAM usage.

Dampfinchen commented:
@compilade I can confirm the issue I've had has been fixed. Good work!

compilade commented on examples/perplexity/perplexity.cpp:
```diff
@@ -1022,6 +1038,38 @@ static std::vector<winogrande_entry> load_winogrande_from_csv(const std::string&
         auto choice2 = line.substr(comma_pos[2]+1, comma_pos[3] - comma_pos[2] - 1);
         auto answer  = line.substr(comma_pos[3]+1, line.size() - comma_pos[3] - 1);
         auto index = line.substr(0, comma_pos[0]);
+        if ('a' <= sentence[0] && sentence[0] <= 'z') {
+            // make the first letter a capital letter
+            sentence[0] -= 'a' - 'A';
+        }
+        for (int i = 0; i < (int) sentence.size() - 1; ++i) {
+            // trim repeated spaces and spaces before punctuation
+            if (sentence[i] == ' ') {
+                char next = sentence[i+1];
+                if (next == ' ' || next == ',' || next == '.' || next == '\'') {
+                    char r[2] = { next, 0 };
+                    sentence.replace(i, 2, r);
+                    --i; // stay at the same index for repeated spaces
+                }
+            } else if (sentence[i] == ',' || sentence[i] == '.') {
+                if (sentence[i] == sentence[i+1]) {
+                    // trim repeated punctuation (forward to work at the end of sentences)
+                    char r[2] = { sentence[i], 0 };
+                    sentence.replace(i, 2, r);
+                    --i; // same index to then run the other checks on that punctuation
+                } else if (0 < i && sentence[i-1] == sentence[i]) {
+                    // trim repeated punctuation (looks back to work with the space trim)
+                    char r[2] = { sentence[i], 0 };
+                    sentence.replace(i-1, 2, r);
+                    i -= 2; // go back because content was shifted
+                } else if (sentence[i+1] != ' ') {
+                    // add missing space after punctuation
+                    // (since the loop stops before the end, this adds no trailing space)
+                    char r[3] = { sentence[i], ' ', 0 };
+                    sentence.replace(i, 1, r);
+                }
+            }
+        }
```
It's somewhat related, because I've changed how the Winogrande score is calculated (it didn't use the correct logits for the choice word of the second choice before, and it used a choice-word-skipping logic to mitigate the effects), and normalizing the sentences helped to make Phi-2_Q4_K_M's score a bit better (`72.6125 +/- 1.2533` when normalizing vs `72.0600 +/- 1.2611` when not normalizing). Some of the sentences in the Winogrande dataset I used has weird punctuation and inconsistent spacing, which sometimes slightly skews the log-likelyhood, but this only affects 111/1267 of the tasks in <https://huggingface.co/datasets/ikawrakow/winogrande-eval-for-llama.cpp>.

```console
$ grep -E "(^[a-z]|\.\w|\.\.| '| \.| ,|  +)" winogrande-debiased-eval.csv | wc -l
111
```

I could try extracting this into another PR or removing it altogether if it's too out-of-place, like if it's better to keep the dataset as-is even if some of the formatting confuses some models.

slaren replied:
I can review the changes in llama.cpp and ggml, but I cannot review these changes. Since `perplexity` is used to test the correctness of many other things, I think it is important that the changes are carefully reviewed. So as it is, either someone else will have to review this PR, or it will have to be split.

compilade replied:
In this case I could make Winogrande work as it does on `master`, even though it's wrong and needlessly complicated, and then fix it in a separate PR. All of the benchmarks in `perplexity` (except Winogrande, but I could fix this) should give the exact same results in this PR as on `master`, which *should* make it relatively easy to test that it works as expected or know if I made a mistake somewhere.

ikawrakow replied:
"Wrong and needlessly complicated": this basically necessitates a separate PR.

slaren replied:
I have been testing hellaswag since there are some changes, and over 400 tasks with I see a difference of 0.25 compared to master (75.00 vs 74.75) with llama 7B F16. Over mistral F16 there is also a small difference (81.25 vs 81.00). Over the full 10042 tasks the result is basically the same so it is probably nothing, but I wonder if something changed that could result in a different computation? I thought it could be due to different matrix sizes, but there is also a difference with MMQ.

compilade replied:
> I wonder if something changed that could result in a different computation?

I've tested with `stories260K.gguf` while showing the logprobs of the 4 choices, and I see that the numbers (compared to `master`) are exactly the same for the first choice, while for the other three, the 4 most significant digits match while the rest seem to vary. So if this changes the result, the choices were already very close to each other. This is either due to some numerical instability in floats or there's something I did wrong. (apparently, it's both)

Okay, while typing this I ran some tests, and I don't understand why, but evaluating the last tokens of the sequences (even without extracting the logits) is apparently necessary to get the exact same numbers. Previously, I simply skipped the evaluation of the last tokens because their output isn't necessary at all. It's weird because the choices are in separate sequences, so the KV cache shouldn't be affected... (tested with Mamba too, and at least my intuition holds there, so the way the KV cache is updated for Transformer-based models makes this different. Why would the KV cells of unrelated sequences be changed slightly like this?) I'll need to run tests for the other benchmarks and will push a fix tomorrow.

ggerganov replied:
See this comment (and the one referenced in it) for explanation of why the results are different: https://github.com/ggerganov/whisper.cpp/issues/1941#issuecomment-1986923227

compilade committed:
perplexity : make Winogrande work as it does on master

The problems with the Winogrande implementation will
need to be fixed in a separate PR to ease review.

compilade committed:
llama : clearer error messages for invalid logits or embeddings ids

* llama : assert all models that can have inp_out_ids

Since the graph topology is now constant, this presence check
can be done even when there are no outputs.

* llama : assert logits and embd buffers exist before writing to them

slaren commented on llama.cpp:
```diff
@@ -8954,6 +9165,70 @@ static void llama_set_inputs(llama_context & lctx, const llama_batch & batch) {
+
+        lctx.buf_output = ggml_backend_buft_alloc_buffer(llama_default_buffer_type_cpu(true), new_size);
+        if (lctx.buf_output == nullptr) {
+            throw std::runtime_error(format("failed to allocate output buffer of size %.2f MiB", new_size / (1024.0 * 1024.0)));
```
This exception is never caught in `llama_decode_internal`, so it will propagate to the user. I think this can be replaced with an error code, all the places where this function is called can handle directly the error.

compilade committed:
llama : handle errors from llama_output_reserve at call sites

compilade committed:
perplexity : make hellaswag and multiple-choice outputs identical to master

Due to how the KV cache is updated, the logprobs for tokens in a batch
are very slightly affected by the other tokens present in the batch,
so to make hellaswag and multiple-choice return exactly the same results
as on master, the last token of each sequence needs to be evaluated
even though its output is not used at all.

This will probably be changed back in the future to make these benchmarks
a tiny bit faster.

* perplexity : fix division by zero when using less than 100 multiple-choice tasks

slaren commented:
@0cc4m would it very hard to add F32 support for `GGML_OP_GET_ROWS` in the Vulkan backend? It will be required after this is merged.

0cc4m commented:
> @0cc4m would it very hard to add F32 support for `GGML_OP_GET_ROWS` in the Vulkan backend? It will be required after this is merged.

Probably not hard, but I'm not confident get_rows works at all right now, which is why I didn't mark it as supported in `ggml_backend_vk_supports_op`. I can take a look this weekend and if it's not too much work add that to #6155

compilade committed:
Merge branch 'master' into compilade/smaller-output-buffer

Notably includes the new repetition penalty default, support for grok-1,
and support for split GGUF.

slaren reviewed (approved):
The only blocker for this PR is the lack of support of get_rows in the Vulkan backend. It is not clear when this will be implemented and we cannot delay this infinitely, so I think this should be merged before more conflicts arise.

compilade committed:
llama : allow loading state saved with a different ctx size

When loading a session file, the context size is now only required to be
at least enough to load the KV cells contained in that session file,
instead of requiring to use exactly the same context size as when saving.

Doing this enables the use-case of extending or shrinking the context size
of a saved session.

This breaks existing session files because the meaning of kv_buf_size
is slightly changed (previously it was the size of the whole KV cache,
now it's only the size of the saved part of it). This allows for
finer-grained sanity checks when loading in an effort to keep kv_buf_size
useful even when the kv_size is changed.

0cc4m commented:
> The only blocker for this PR is the lack of support of get_rows in the Vulkan backend. It is not clear when this will be implemented and we cannot delay this infinitely, so I think this should be merged before more conflicts arise.

I've got that mostly done this weekend, but didn't have the time to figure out the last bugs in the implementation.

ggerganov committed:
llama : minor

ggml-ci

ggerganov reviewed (approved):
Let's merge and finish the Vulkan kernels from `master`. @0cc4m if it takes longer to resolve the issues, we can put a notice in the README with the last stable version (i.e. before this PR) until it is ready

compilade committed:
readme : update recent API changes, and warn about Vulkan

ikawrakow commented:
This PR breaks MoE models. Well, I don't know if all MoE models, but for sure Mixtral8x7B.

Everything is fine until 55c1b2a3bbd470e9e2a3a0618b92cf64a885f806. Then
```
git co 557410b8f06380560155ac7fcb8316d71ddc9837 && make -j

./perplexity -m some_model -f some_test -t 1 -ngl 100
main: build = 2540 (557410b8)
main: built with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu
main: seed  = 1711546930
llama_model_loader: loaded meta data with 25 key-value pairs and 995 tensors from jun2.bin (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = hf
llama_model_loader: - kv   2:                       llama.context_length u32              = 32768
llama_model_loader: - kv   3:                     llama.embedding_length u32              = 4096
llama_model_loader: - kv   4:                          llama.block_count u32              = 32
llama_model_loader: - kv   5:                  llama.feed_forward_length u32              = 14336
llama_model_loader: - kv   6:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv   7:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv   8:              llama.attention.head_count_kv u32              = 8
llama_model_loader: - kv   9:                         llama.expert_count u32              = 8
llama_model_loader: - kv  10:                    llama.expert_used_count u32              = 2
llama_model_loader: - kv  11:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  12:                       llama.rope.freq_base f32              = 1000000.000000
llama_model_loader: - kv  13:                          general.file_type u32              = 24
llama_model_loader: - kv  14:                       tokenizer.ggml.model str              = llama
llama_model_loader: - kv  15:                      tokenizer.ggml.tokens arr[str,32000]   = ["<unk>", "<s>", "</s>", "<0x00>", "<...
llama_model_loader: - kv  16:                      tokenizer.ggml.scores arr[f32,32000]   = [0.000000, 0.000000, 0.000000, 0.0000...
llama_model_loader: - kv  17:                  tokenizer.ggml.token_type arr[i32,32000]   = [2, 3, 3, 6, 6, 6, 6, 6, 6, 6, 6, 6, ...
llama_model_loader: - kv  18:                      tokenizer.ggml.merges arr[str,58980]   = ["▁ t", "i n", "e r", "▁ a", "h e...
llama_model_loader: - kv  19:                tokenizer.ggml.bos_token_id u32              = 1
llama_model_loader: - kv  20:                tokenizer.ggml.eos_token_id u32              = 2
llama_model_loader: - kv  21:            tokenizer.ggml.unknown_token_id u32              = 0
llama_model_loader: - kv  22:               tokenizer.ggml.add_bos_token bool             = true
llama_model_loader: - kv  23:               tokenizer.ggml.add_eos_token bool             = false
llama_model_loader: - kv  24:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:   65 tensors
llama_model_loader: - type  f16:   32 tensors
llama_model_loader: - type q2_K:   33 tensors
llama_model_loader: - type q4_K:   64 tensors
llama_model_loader: - type q5_K:   33 tensors
llama_model_loader: - type iq1_s:  768 tensors
llm_load_vocab: special tokens definition check successful ( 259/32000 ).
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = llama
llm_load_print_meta: vocab type       = SPM
llm_load_print_meta: n_vocab          = 32000
llm_load_print_meta: n_merges         = 0
llm_load_print_meta: n_ctx_train      = 32768
llm_load_print_meta: n_embd           = 4096
llm_load_print_meta: n_head           = 32
llm_load_print_meta: n_head_kv        = 8
llm_load_print_meta: n_layer          = 32
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_embd_head_k    = 128
llm_load_print_meta: n_embd_head_v    = 128
llm_load_print_meta: n_gqa            = 4
llm_load_print_meta: n_embd_k_gqa     = 1024
llm_load_print_meta: n_embd_v_gqa     = 1024
llm_load_print_meta: f_norm_eps       = 0.0e+00
llm_load_print_meta: f_norm_rms_eps   = 1.0e-05
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: f_logit_scale    = 0.0e+00
llm_load_print_meta: n_ff             = 14336
llm_load_print_meta: n_expert         = 8
llm_load_print_meta: n_expert_used    = 2
llm_load_print_meta: causal attn      = 1
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = 0
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 1000000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_yarn_orig_ctx  = 32768
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: ssm_d_conv       = 0
llm_load_print_meta: ssm_d_inner      = 0
llm_load_print_meta: ssm_d_state      = 0
llm_load_print_meta: ssm_dt_rank      = 0
llm_load_print_meta: model type       = 7B
llm_load_print_meta: model ftype      = IQ1_S - 1.5625 bpw
llm_load_print_meta: model params     = 46.70 B
llm_load_print_meta: model size       = 9.14 GiB (1.68 BPW) 
llm_load_print_meta: general.name     = hf
llm_load_print_meta: BOS token        = 1 '<s>'
llm_load_print_meta: EOS token        = 2 '</s>'
llm_load_print_meta: UNK token        = 0 '<unk>'
llm_load_print_meta: LF token         = 13 '<0x0A>'
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:   no
ggml_cuda_init: CUDA_USE_TENSOR_CORES: yes
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 4080, compute capability 8.9, VMM: yes
llm_load_tensors: ggml ctx size =    0.76 MiB
llm_load_tensors: offloading 32 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 33/33 layers to GPU
llm_load_tensors:        CPU buffer size =    41.02 MiB
llm_load_tensors:      CUDA0 buffer size =  9322.95 MiB
....................................................................................................
llama_new_context_with_model: n_ctx      = 4096
llama_new_context_with_model: n_batch    = 4096
llama_new_context_with_model: n_ubatch   = 512
llama_new_context_with_model: freq_base  = 1000000.0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init:      CUDA0 KV buffer size =   512.00 MiB
llama_new_context_with_model: KV self size  =  512.00 MiB, K (f16):  256.00 MiB, V (f16):  256.00 MiB
llama_new_context_with_model:  CUDA_Host  output buffer size =     0.49 MiB
llama_new_context_with_model:      CUDA0 compute buffer size =   316.04 MiB
llama_new_context_with_model:  CUDA_Host compute buffer size =    16.01 MiB
llama_new_context_with_model: graph nodes  = 1670
llama_new_context_with_model: graph splits = 2

system_info: n_threads = 1 / 64 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | MATMUL_INT8 = 0 | 
perplexity: tokenizing the input ..
perplexity: tokenization took 658.483 ms
perplexity: calculating perplexity over 80 chunks, n_ctx=4096, batch_size=4096, n_seq=1
GGML_ASSERT: /home/iwan/other/llama.cpp/ggml.c:2941: view_src == NULL || data_size + view_offs <= ggml_nbytes(view_src)
Could not attach to process.  If your uid matches the uid of the target
process, check the setting of /proc/sys/kernel/yama/ptrace_scope, or try
again as the root user.  For more details, see /etc/sysctl.d/10-ptrace.conf
ptrace: Operation not permitted.
No stack.
The program is not being run.
Aborted (core dumped)
```

Wuzzooy commented:
command-r seems broken, it works until build b2536.
I'm getting "GGML_ASSERT: D:\a\llama.cpp\llama.cpp\ggml.c:3648: ggml_can_repeat(b, a)"
``INFO [     server_params_parse] logging to file is disabled. | tid="196984" timestamp=1711615937
INFO [                    main] build info | tid="196984" timestamp=1711615937 build=2554 commit="25f4a613"
INFO [                    main] system info | tid="196984" timestamp=1711615937 n_threads=7 n_threads_batch=-1 total_threads=8 system_info="AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 0 | VSX = 0 | MATMUL_INT8 = 0 | "
llama_model_loader: loaded meta data with 23 key-value pairs and 322 tensors from G:\AI\models\c4ai-command-r-v01-Q4_K_S.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = command-r
llama_model_loader: - kv   1:                               general.name str              = c4ai-command-r-v01
llama_model_loader: - kv   2:                      command-r.block_count u32              = 40
llama_model_loader: - kv   3:                   command-r.context_length u32              = 131072
llama_model_loader: - kv   4:                 command-r.embedding_length u32              = 8192
llama_model_loader: - kv   5:              command-r.feed_forward_length u32              = 22528
llama_model_loader: - kv   6:             command-r.attention.head_count u32              = 64
llama_model_loader: - kv   7:          command-r.attention.head_count_kv u32              = 64
llama_model_loader: - kv   8:                   command-r.rope.freq_base f32              = 8000000.000000
llama_model_loader: - kv   9:     command-r.attention.layer_norm_epsilon f32              = 0.000010
llama_model_loader: - kv  10:                          general.file_type u32              = 14
llama_model_loader: - kv  11:                      command-r.logit_scale f32              = 0.062500
llama_model_loader: - kv  12:                command-r.rope.scaling.type str              = none
llama_model_loader: - kv  13:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  14:                      tokenizer.ggml.tokens arr[str,256000]  = ["<PAD>", "<UNK>", "<CLS>", "<SEP>", ...
llama_model_loader: - kv  15:                  tokenizer.ggml.token_type arr[i32,256000]  = [3, 3, 3, 3, 3, 3, 3, 3, 1, 1, 1, 1, ...
llama_model_loader: - kv  16:                      tokenizer.ggml.merges arr[str,253333]  = ["─á ─á", "─á t", "e r", "i n", "─á a...
llama_model_loader: - kv  17:                tokenizer.ggml.bos_token_id u32              = 5
llama_model_loader: - kv  18:                tokenizer.ggml.eos_token_id u32              = 255001
llama_model_loader: - kv  19:            tokenizer.ggml.padding_token_id u32              = 0
llama_model_loader: - kv  20:               tokenizer.ggml.add_bos_token bool             = true
llama_model_loader: - kv  21:               tokenizer.ggml.add_eos_token bool             = false
llama_model_loader: - kv  22:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:   41 tensors
llama_model_loader: - type q4_K:  271 tensors
llama_model_loader: - type q5_K:    9 tensors
llama_model_loader: - type q6_K:    1 tensors
llm_load_vocab: special tokens definition check successful ( 1008/256000 ).
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = command-r
llm_load_print_meta: vocab type       = BPE
llm_load_print_meta: n_vocab          = 256000
llm_load_print_meta: n_merges         = 253333
llm_load_print_meta: n_ctx_train      = 131072
llm_load_print_meta: n_embd           = 8192
llm_load_print_meta: n_head           = 64
llm_load_print_meta: n_head_kv        = 64
llm_load_print_meta: n_layer          = 40
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_embd_head_k    = 128
llm_load_print_meta: n_embd_head_v    = 128
llm_load_print_meta: n_gqa            = 1
llm_load_print_meta: n_embd_k_gqa     = 8192
llm_load_print_meta: n_embd_v_gqa     = 8192
llm_load_print_meta: f_norm_eps       = 1.0e-05
llm_load_print_meta: f_norm_rms_eps   = 0.0e+00
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: f_logit_scale    = 6.2e-02
llm_load_print_meta: n_ff             = 22528
llm_load_print_meta: n_expert         = 0
llm_load_print_meta: n_expert_used    = 0
llm_load_print_meta: causal attn      = 1
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = 0
llm_load_print_meta: rope scaling     = none
llm_load_print_meta: freq_base_train  = 8000000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_yarn_orig_ctx  = 131072
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: ssm_d_conv       = 0
llm_load_print_meta: ssm_d_inner      = 0
llm_load_print_meta: ssm_d_state      = 0
llm_load_print_meta: ssm_dt_rank      = 0
llm_load_print_meta: model type       = 35B
llm_load_print_meta: model ftype      = Q4_K - Small
llm_load_print_meta: model params     = 34.98 B
llm_load_print_meta: model size       = 18.97 GiB (4.66 BPW)
llm_load_print_meta: general.name     = c4ai-command-r-v01
llm_load_print_meta: BOS token        = 5 '<BOS_TOKEN>'
llm_load_print_meta: EOS token        = 255001 '<|END_OF_TURN_TOKEN|>'
llm_load_print_meta: PAD token        = 0 '<PAD>'
llm_load_print_meta: LF token         = 136 '├ä'
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:   no
ggml_cuda_init: CUDA_USE_TENSOR_CORES: yes
ggml_cuda_init: found 2 CUDA devices:
  Device 0: NVIDIA GeForce RTX 4070 Ti, compute capability 8.9, VMM: yes
  Device 1: NVIDIA GeForce RTX 3090 Ti, compute capability 8.6, VMM: yes
llm_load_tensors: ggml ctx size =    0.37 MiB
llm_load_tensors: offloading 40 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 41/41 layers to GPU
llm_load_tensors:        CPU buffer size =  1640.62 MiB
llm_load_tensors:      CUDA0 buffer size =  5434.38 MiB
llm_load_tensors:      CUDA1 buffer size = 13989.53 MiB
.......................................................................................
llama_new_context_with_model: n_ctx      = 4096
llama_new_context_with_model: n_batch    = 2048
llama_new_context_with_model: n_ubatch   = 512
llama_new_context_with_model: freq_base  = 8000000.0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init:      CUDA0 KV buffer size =  1536.00 MiB
llama_kv_cache_init:      CUDA1 KV buffer size =  3584.00 MiB
llama_new_context_with_model: KV self size  = 5120.00 MiB, K (f16): 2560.00 MiB, V (f16): 2560.00 MiB
llama_new_context_with_model:  CUDA_Host  output buffer size =     1.95 MiB
llama_new_context_with_model: pipeline parallelism enabled (n_copies=4)
llama_new_context_with_model:      CUDA0 compute buffer size =   672.01 MiB
llama_new_context_with_model:      CUDA1 compute buffer size =   672.02 MiB
llama_new_context_with_model:  CUDA_Host compute buffer size =    48.02 MiB
llama_new_context_with_model: graph nodes  = 1247
llama_new_context_with_model: graph splits = 3
[1711615946] warming up the model with an empty run
INFO [                    init] initializing slots | tid="196984" timestamp=1711615946 n_slots=1
INFO [                    init] new slot | tid="196984" timestamp=1711615946 id_slot=0 n_ctx_slot=4096
INFO [                    main] model loaded | tid="196984" timestamp=1711615946
INFO [                    main] chat template | tid="196984" timestamp=1711615946 chat_example="<|im_start|>system\nYou are a helpful assistant<|im_end|>\n<|im_start|>user\nHello<|im_end|>\n<|im_start|>assistant\nHi there<|im_end|>\n<|im_start|>user\nHow are you?<|im_end|>\n<|im_start|>assistant\n" built_in=true
INFO [                    main] HTTP server listening | tid="196984" timestamp=1711615946 hostname="127.0.0.1" port="8080" n_threads_http="7"
INFO [            update_slots] all slots are idle | tid="196984" timestamp=1711615946
INFO [   launch_slot_with_task] slot is processing task | tid="196984" timestamp=1711615976 id_slot=0 id_task=0
INFO [            update_slots] kv cache rm [p0, end) | tid="196984" timestamp=1711615976 id_slot=0 id_task=0 p0=0
GGML_ASSERT: D:\a\llama.cpp\llama.cpp\ggml.c:3648: ggml_can_repeat(b, a)

compilade commented:
> command-r seems broken, it works until build b2536.
> I'm getting "GGML_ASSERT: D:\a\llama.cpp\llama.cpp\ggml.c:3648: ggml_can_repeat(b, a)"

@Wuzzooy I'm truly sorry about that, it seems the model's graph was structured a bit differently than the other models (`inpSA` was named `ffn_inp` instead), and I overlooked a tensor which should have gone through the new `inp_out_ids` dance.

Should be fixed by #6367 (hopefully).
